File: src/main/java/me/kuku/yuq/job/HostLocJob.java
package me.kuku.yuq.job;

import com.IceCreamQAQ.Yu.annotation.Cron;
import com.IceCreamQAQ.Yu.annotation.JobCenter;
import me.kuku.yuq.entity.*;
import me.kuku.yuq.logic.HostLocLogic;
import me.kuku.yuq.utils.BotUtils;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@JobCenter
public class HostLocJob {

	@Inject
	private HostLocService hostLocService;
	@Inject
	private HostLocLogic hostLocLogic;
	@Inject
	private GroupService groupService;
	@Inject
	private QqService qqService;

	private int locId = 0;

	@Cron("At::d::07:20:13")
	public void sign(){
		List<HostLocEntity> list = hostLocService.findBySign(true);
		for (HostLocEntity hostLocEntity : list) {
			try {
				String cookie = hostLocEntity.getCookie();
				boolean isLogin = hostLocLogic.isLogin(cookie);
				if (isLogin) hostLocLogic.sign(cookie);
				else {
					QqEntity qqEntity = hostLocEntity.getQqEntity();
					BotUtils.sendMessage(qqEntity,
							"loc签到失败，cookie已失效，请重新绑定，并在群中发送<loc签到>进行手动签到！");
					hostLocService.delete(hostLocEntity);
				}
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}

	@Cron("1m")
	public void locMonitor() {
		List<Map<String, String>> list = hostLocLogic.post();
		if (list.size() == 0) return;
		List<Map<String, String>> newList = new ArrayList<>();
		if (locId != 0){
			for (Map<String, String> map : list) {
				if (Integer.parseInt(map.get("id")) <= locId) break;
				newList.add(map);
			}
		}
		locId = Integer.parseInt(list.get(0).get("id"));
		List<HostLocEntity> qqList = hostLocService.findByMonitor(true);
		for (Map<String, String> map : newList) {
			String str = "Loc有新帖了！！" + "\n" +
					"标题：" + map.get("title") + "\n" +
					"昵称：" + map.get("name") + "\n" +
					"链接：" + map.get("url");
			for (HostLocEntity hostLocEntity : qqList) {
				QqEntity qqEntity = hostLocEntity.getQqEntity();
				BotUtils.sendMessage(qqEntity, str);
			}
		}
	}

}

File: src/main/kotlin/me/kuku/yuq/job/HostLocJob.kt
package me.kuku.yuq.job

import com.IceCreamQAQ.Yu.annotation.Cron
import com.IceCreamQAQ.Yu.annotation.JobCenter
import kotlinx.coroutines.delay
import me.kuku.yuq.entity.HostLocService
import me.kuku.yuq.entity.Status
import me.kuku.yuq.logic.HostLocLogic
import me.kuku.yuq.logic.HostLocPost
import me.kuku.yuq.utils.YuqUtils
import javax.inject.Inject

@JobCenter
class HostLocJob @Inject constructor(
    private val hostLocService: HostLocService
) {

    private var locId = 0

    @Cron("1s")
    suspend fun locPush() {
        val list = HostLocLogic.post()
        if (list.isEmpty()) return
        val newList = mutableListOf<HostLocPost>()
        if (locId != 0) {
            for (hostLocPost in list) {
                if (hostLocPost.id <= locId) break
                newList.add(hostLocPost)
            }
        }
        locId = list[0].id
        for (hostLocPost in newList) {
            delay(3000)
            val hostLocList = hostLocService.findByStatus(Status.ON).filter { it.config.push == Status.ON }
            for (hostLocEntity in hostLocList) {
                val str = """
                    Loc有新帖了！！
                    标题：${hostLocPost.title}
                    昵称：${hostLocPost.name}
                    链接：${hostLocPost.url}
                    内容：${HostLocLogic.postContent(hostLocPost.url, hostLocEntity.cookie)}
                """.trimIndent()
                YuqUtils.sendMessage(hostLocEntity.qqEntity!!, str)
            }
        }
    }

    @Cron("04:12")
    fun sign() {
        val list = hostLocService.findByStatus(Status.ON).filter { it.config.sign == Status.ON }
        for (hostLocEntity in list) {
            HostLocLogic.sign(hostLocEntity.cookie)
        }
    }



}